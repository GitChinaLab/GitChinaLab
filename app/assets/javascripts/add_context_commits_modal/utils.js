export const findCommitIndex = (commits, commitShortId) => {
  return commits.findIndex((commit) => commit.short_id === commitShortId);
};

export const setCommitStatus = (commits, commitIndex, selected) => {
  const tempCommits = [...commits];
  tempCommits[commitIndex] = {
    ...tempCommits[commitIndex],
    isSelected: selected,
  };
  return tempCommits;
};

export const removeIfReadyToBeRemoved = (toRemoveCommits, commitShortId) => {
  const tempToRemoveCommits = [...toRemoveCommits];
  const isPresentInToRemove = tempToRemoveCommits.indexOf(commitShortId);
  if (isPresentInToRemove !== -1) {
    tempToRemoveCommits.splice(isPresentInToRemove, 1);
  }

  return tempToRemoveCommits;
};

export const removeIfPresent = (selectedCommits, commitShortId) => {
  const tempSelectedCommits = [...selectedCommits];
  const selectedCommitsIndex = findCommitIndex(tempSelectedCommits, commitShortId);
  if (selectedCommitsIndex !== -1) {
    tempSelectedCommits.splice(selectedCommitsIndex, 1);
  }

  return tempSelectedCommits;
};
