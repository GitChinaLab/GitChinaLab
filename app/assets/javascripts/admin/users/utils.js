export const generateUserPaths = (paths, id) => {
  return Object.fromEntries(
    Object.entries(paths).map(([action, genericPath]) => {
      return [action, genericPath.replace('id', id)];
    }),
  );
};
