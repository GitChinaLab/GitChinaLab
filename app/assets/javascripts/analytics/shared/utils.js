import dateFormat from 'dateformat';
import { urlQueryToFilter } from '~/vue_shared/components/filtered_search_bar/filtered_search_utils';
import { dateFormats } from './constants';

export const filterBySearchTerm = (data = [], searchTerm = '', filterByKey = 'name') => {
  if (!searchTerm?.length) return data;
  return data.filter((item) => item[filterByKey].toLowerCase().includes(searchTerm.toLowerCase()));
};

export const toYmd = (date) => dateFormat(date, dateFormats.isoDate);

/**
 * Takes a url and extracts query parameters used for the shared
 * filter bar
 *
 * @param {string} url The URL to extract query parameters from
 * @returns {Object}
 */
export const extractFilterQueryParameters = (url = '') => {
  const {
    source_branch_name = null,
    target_branch_name = null,
    author_username = null,
    milestone_title = null,
    assignee_username = [],
    label_name = [],
  } = urlQueryToFilter(url);

  return {
    selectedSourceBranch: source_branch_name,
    selectedTargetBranch: target_branch_name,
    selectedAuthor: author_username,
    selectedMilestone: milestone_title,
    selectedAssigneeList: assignee_username,
    selectedLabelList: label_name,
  };
};

/**
 * Takes a url and extracts sorting and pagination query parameters into an object
 *
 * @param {string} url The URL to extract query parameters from
 * @returns {Object}
 */
export const extractPaginationQueryParameters = (url = '') => {
  const { sort, direction, page } = urlQueryToFilter(url);
  return {
    sort: sort?.value || null,
    direction: direction?.value || null,
    page: page?.value || null,
  };
};

export const getDataZoomOption = ({
  totalItems = 0,
  maxItemsPerPage = 40,
  dataZoom = [{ type: 'slider', bottom: 10, start: 0 }],
}) => {
  if (totalItems <= maxItemsPerPage) {
    return {};
  }

  const intervalEnd = Math.ceil((maxItemsPerPage / totalItems) * 100);

  return dataZoom.map((item) => {
    return {
      ...item,
      end: intervalEnd,
    };
  });
};
