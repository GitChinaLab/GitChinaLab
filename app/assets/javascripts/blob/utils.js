import Editor from '~/editor/source_editor';

export function initSourceEditor({ el, ...args }) {
  const editor = new Editor({
    scrollbar: {
      alwaysConsumeMouseWheel: false,
    },
  });

  return editor.createInstance({
    el,
    ...args,
  });
}

export default () => ({});
