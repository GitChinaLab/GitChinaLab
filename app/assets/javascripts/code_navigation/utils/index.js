export const cachedData = new Map();

export const getCurrentHoverElement = () => cachedData.get('current');
export const setCurrentHoverElement = (el) => cachedData.set('current', el);

export const addInteractionClass = (path, d) => {
  const lineNumber = d.start_line + 1;
  const lines = document
    .querySelector(`[data-path="${path}"]`)
    .querySelectorAll(`.blob-content #LC${lineNumber}, .line_content:not(.old) #LC${lineNumber}`);
  if (!lines?.length) return;

  lines.forEach((line) => {
    let charCount = 0;
    const el = [...line.childNodes].find(({ textContent }) => {
      if (charCount === d.start_char) return true;
      charCount += textContent.length;
      return false;
    });

    if (el) {
      el.setAttribute('data-char-index', d.start_char);
      el.setAttribute('data-line-index', d.start_line);
      el.classList.add('cursor-pointer', 'code-navigation', 'js-code-navigation');
      el.closest('.line').classList.add('code-navigation-line');
    }
  });
};
