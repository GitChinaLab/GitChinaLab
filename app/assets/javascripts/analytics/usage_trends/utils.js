import { masks } from 'dateformat';
import { get } from 'lodash';
import { formatDate } from '~/lib/utils/datetime_utility';

const { isoDate } = masks;

/**
 * Takes an array of items and returns one item per month with the average of the `count`s from that month
 * @param  {Array} items
 * @param  {Number} items[index].count value to be averaged
 * @param  {String} items[index].recordedAt item dateTime time stamp to be collected into a month
 * @param  {Object} options
 * @param  {Object} options.shouldRound an option to specify whether the retuned averages should be rounded
 * @return {Array} items collected into [month, average],
 * where month is a dateTime string representing the first of the given month
 * and average is the average of the count
 */
export function getAverageByMonth(items = [], options = {}) {
  const { shouldRound = false } = options;
  const itemsMap = items.reduce((memo, item) => {
    const { count, recordedAt } = item;
    const date = new Date(recordedAt);
    const month = formatDate(new Date(date.getFullYear(), date.getMonth(), 1), isoDate);
    if (memo[month]) {
      const { sum, recordCount } = memo[month];
      return { ...memo, [month]: { sum: sum + count, recordCount: recordCount + 1 } };
    }

    return { ...memo, [month]: { sum: count, recordCount: 1 } };
  }, {});

  return Object.keys(itemsMap).map((month) => {
    const { sum, recordCount } = itemsMap[month];
    const avg = sum / recordCount;
    if (shouldRound) {
      return [month, Math.round(avg)];
    }

    return [month, avg];
  });
}

/**
 * Takes an array of usage counts and returns the last item in the list
 * @param  {Array} arr array of usage counts in the form { count: Number, recordedAt: date String }
 * @return {String} the 'recordedAt' value of the earliest item
 */
export const getEarliestDate = (arr = []) => {
  const len = arr.length;
  return get(arr, `[${len - 1}].recordedAt`, null);
};

/**
 * Takes an array of queries and produces an object with the query identifier as key
 * and a supplied defaultValue as its value
 * @param  {Array} queries array of chart query configs,
 *                 see ./analytics/usage_trends/components/charts_config.js
 * @param  {any}   defaultValue value to set each identifier to
 * @return {Object} key value pair of the form { queryIdentifier: defaultValue }
 */
export const generateDataKeys = (queries, defaultValue) =>
  queries.reduce(
    (acc, { identifier }) => ({
      ...acc,
      [identifier]: defaultValue,
    }),
    {},
  );
