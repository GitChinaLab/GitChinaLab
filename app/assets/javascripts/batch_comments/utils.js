import { getFormData } from '~/diffs/store/utils';

export const getDraftReplyFormData = (data) => ({
  endpoint: data.notesData.draftsPath,
  data,
});

export const getDraftFormData = (params) => ({
  endpoint: params.notesData.draftsPath,
  data: getFormData(params),
});

export const parallelLineKey = (line, side) => (line[side] ? line[side].line_code : '');

export const showDraftOnSide = (line, side) => {
  // inline mode
  if (side === null) {
    return true;
  }

  // parallel
  if (side === 'left' || side === 'right') {
    const otherSide = side === 'left' ? 'right' : 'left';
    const thisCode = (line[side] && line[side].line_code) || '';
    const otherCode = (line[otherSide] && line[otherSide].line_code) || '';

    // either the lineCodes are different
    // or if they're the same, only show on the left side
    if (thisCode !== otherCode || (side === 'left' && thisCode === otherCode)) {
      return true;
    }
  }

  return false;
};
