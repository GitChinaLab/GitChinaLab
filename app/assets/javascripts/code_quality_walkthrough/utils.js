import { TRACKING_CONTEXT_SCHEMA } from '~/experimentation/constants';
import { getExperimentData } from '~/experimentation/utils';
import { setCookie, getCookie } from '~/lib/utils/common_utils';
import { getParameterByName } from '~/lib/utils/url_utility';
import Tracking from '~/tracking';
import { EXPERIMENT_NAME } from './constants';

export function getExperimentSettings() {
  return JSON.parse(getCookie(EXPERIMENT_NAME) || '{}');
}

export function setExperimentSettings(settings) {
  setCookie(EXPERIMENT_NAME, settings);
}

export function isWalkthroughEnabled() {
  return getParameterByName(EXPERIMENT_NAME);
}

export function track(action) {
  const { data } = getExperimentSettings();

  if (data) {
    Tracking.event(EXPERIMENT_NAME, action, {
      context: {
        schema: TRACKING_CONTEXT_SCHEMA,
        data,
      },
    });
  }
}

export function startCodeQualityWalkthrough() {
  const data = getExperimentData(EXPERIMENT_NAME);

  if (data) {
    setExperimentSettings({ data });
  }
}
