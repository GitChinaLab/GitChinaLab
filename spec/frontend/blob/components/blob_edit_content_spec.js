import { shallowMount } from '@vue/test-utils';
import { nextTick } from 'vue';
import BlobEditContent from '~/blob/components/blob_edit_content.vue';
import * as utils from '~/blob/utils';

jest.mock('~/editor/source_editor');

describe('Blob Header Editing', () => {
  let wrapper;
  const onDidChangeModelContent = jest.fn();
  const updateModelLanguage = jest.fn();
  const getValue = jest.fn();
  const value = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.';
  const fileName = 'lorem.txt';
  const fileGlobalId = 'snippet_777';

  function createComponent(props = {}) {
    wrapper = shallowMount(BlobEditContent, {
      propsData: {
        value,
        fileName,
        fileGlobalId,
        ...props,
      },
    });
  }

  beforeEach(() => {
    jest.spyOn(utils, 'initSourceEditor').mockImplementation(() => ({
      onDidChangeModelContent,
      updateModelLanguage,
      getValue,
      dispose: jest.fn(),
    }));

    createComponent();
  });

  afterEach(() => {
    wrapper.destroy();
  });

  const triggerChangeContent = (val) => {
    getValue.mockReturnValue(val);
    const [cb] = onDidChangeModelContent.mock.calls[0];

    cb();

    jest.runOnlyPendingTimers();
  };

  describe('rendering', () => {
    it('matches the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('renders content', () => {
      expect(wrapper.text()).toContain(value);
    });
  });

  describe('functionality', () => {
    it('does not fail without content', () => {
      const spy = jest.spyOn(global.console, 'error');
      createComponent({ value: undefined });

      expect(spy).not.toHaveBeenCalled();
      expect(wrapper.find('#editor').exists()).toBe(true);
    });

    it('initialises Source Editor', () => {
      const el = wrapper.find({ ref: 'editor' }).element;
      expect(utils.initSourceEditor).toHaveBeenCalledWith({
        el,
        blobPath: fileName,
        blobGlobalId: fileGlobalId,
        blobContent: value,
      });
    });

    it('reacts to the changes in fileName', () => {
      const newFileName = 'ipsum.txt';

      wrapper.setProps({
        fileName: newFileName,
      });

      return nextTick().then(() => {
        expect(updateModelLanguage).toHaveBeenCalledWith(newFileName);
      });
    });

    it('registers callback with editor onChangeContent', () => {
      expect(onDidChangeModelContent).toHaveBeenCalledWith(expect.any(Function));
    });

    it('emits input event when the blob content is changed', () => {
      expect(wrapper.emitted().input).toBeUndefined();

      triggerChangeContent(value);

      expect(wrapper.emitted().input).toEqual([[value]]);
    });
  });
});
