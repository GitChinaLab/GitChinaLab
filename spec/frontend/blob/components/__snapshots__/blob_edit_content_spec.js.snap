// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blob Header Editing rendering matches the snapshot 1`] = `
<div
  class="file-content code"
>
  <div
    data-editor-loading=""
    id="editor"
  >
    <pre
      class="editor-loading-content"
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit.
    </pre>
  </div>
</div>
`;
